### ISSUE TITLE
ZeroLengthPredicate: gives false warning on File.stat

### ISSUE THREAD

**Issue by perlun** [rubocop-hq/rubocop#2841]
The `ZeroLengthPredicate` is a bit too aggressive in its heuristics. The code below gives a warning:

``` ruby
def manifest_exist(folder)
  manifest_file = File.join(folder, 'app.manifest')

  File.stat(manifest_file).size == 0
end
```

The warning is this:

```
foo.rb:4:3: C: Style/ZeroLengthPredicate: Use empty? instead of size == 0.
  File.stat(manifest_file).size == 0
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

This is a problem since the `File::Stat` class doesn't even _have_ an `empty?` method defined. :smile: 

```
$ irb
2.3.0 :001 > File.stat('Gemfile').empty?
NoMethodError: undefined method `empty?' for #<File::Stat:0x007f86dc902080>
    from (irb):1
    from /Users/plundberg/.rvm/rubies/ruby-2.3.0/bin/irb:11:in `<main>'
```

I can of course easily disable this on the line in question. This issue exists to remind us that the detection code in this cop should be made a bit smarter (if possible), to avoid treating this line as a Rubocop offense.

**Comment by alexdowad** [comment]
> This issue exists to remind us that the detection code in this cop should be made a bit smarter (if possible)

Thanks for the good example. Do you have any ideas how to make the cop smarter?

**Comment by perlun** [comment]
Good question. I looked a bit at the code in https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/style/zero_length_predicate.rb and it's obviously a whole lot more work to do it "properly" here, i.e. trying to check types/etc.

A simple change for now would be to remove the lint for `size == 0` (and `size != 0`) cases, and only run it for `length == 0` etc. cases. I would think that would be the _easy_ way to avoid it reporting false errors.

**Comment by Drenmi** [comment]
You can follow some of the discussion that preceded the implementation of the cop in [this issue](https://github.com/bbatsov/rubocop/issues/2620).

Because of the nature of dynamically typed, object oriented languages, and the use of duck typing in Ruby, it is impossible to avoid false positives. Attempting to do things like type checking (except for literals) would be outside the scope of what's possible with static code analysis.

If I correctly understand the intention of @bbatsov, Rubocop is meant to be tailored to the Ruby standard library, and false positives should be handled through disabling the cop, either inline for a certain piece of code, or altogether.

Within those constraints, however, the cops should obviously be as "smart" as possible, as to minimize the number of duds. :smile:

**Comment by alexdowad** [comment]
> Within those constraints, however, the cops should obviously be as "smart" as possible, as to minimize the number of duds.

Absolutely. If these cops prod someone to make their library follow the conventions of stdlib more closely, that's not a bad thing either.

**Comment by alexdowad** [comment]
So do we want to stop checking for `size == 0` and only look for `length == 0`? @bbatsov?

**Comment by bbatsov** [comment]
I'm fine with making this configurable, but I'm not fine with checking only for `length == 0`.

**Comment by jcoyne** [comment]
I'm also hitting this with checking the size of an uploaded file:

```
```

**Comment by keithbro** [comment]
`StringIO` also.

**Comment by mikegee** [comment]
It seems like a lot to ask, but it would be great if these classes implemented `#empty?`.

**Comment by alexdowad** [comment]
@mikegee You can try to contribute patches to MRI Ruby. They accept PRs through GitHub now.

**Comment by mikegee** [comment]
I’m doing a bad job arguing for `#empty?` at https://bugs.ruby-lang.org/issues/14136  Does anybody else want to contribute to that discussion.

**Comment by bbatsov** [comment]
I've added a few comments myself, but I'm not holding my breath. Generally for this specific problem it seems to me we should discourage people from using `File.stat` to check for file emptiness.

**Comment by jcoyne** [comment]
Something having a `0` size doesn't necessarily imply emptiness unless the thing you have 0 of is the volume of a container. I think this is just over generalization of a solution onto too large of a problem space.  This cop should be considered for removal.

**Comment by Drenmi** [comment]
To get somewhere with this, I created [a patch](https://github.com/rubocop-hq/rubocop/pull/6303) to reduce known false positives that were reported in this thread. I think that's about the extent to what we can go with static analysis. For those who disagree with this cop in general, there's always the option to disable it. 🙂
